### ISSUE TITLE
Tracking Issue for `cmp::minmax{_by,_by_key}`

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#115939]
Feature gate: `#![feature(cmp_minmax)]`

This is a tracking issue for `core::cmp::minmax{_by,_by_key}`.

Those functions take two values and return an array of `[min, max]`. This is essentially a sort specialized for two arguments.

### Public API

```rust
// core::cmp

pub fn minmax<T>(v1: T, v2: T) -> [T; 2]
where
    T: Ord;

pub fn minmax_by<T, F>(v1: T, v2: T, compare: F) -> [T; 2]
where
    F: FnOnce(&T, &T) -> Ordering;

pub fn minmax_by_key<T, F, K>(v1: T, v2: T, f: F) -> [T; 2]
where
    F: FnMut(&T) -> K,
    K: Ord;
```

### Steps / History

- [ ] Implementation: #109409
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by scottmcm** [comment]
Big fan.  `min`/`max` always discard one of the arguments, so they're often useless with owned non-`Copy` things.  This API fixes that by returning both.

**Comment by WaffleLapkin** [comment]
@Jules-Bertholet yes, of course. and the fix would be the same (either using HKT, which is unergonomic in rust, or having a separate function for `F: FnMut(&T) -> &impl ?Sized + Ord`, or adding a bound on function return, such that we are adding a bound on something that is inside the binder)

**Comment by timvermeulen** [comment]
I knew this looked familiar 🙂 https://github.com/rust-lang/rust/pull/73857
Obviously a +1 from me, this concept is useful surprisingly often. A less practical reason why I like this is that it's the only `min`/`max`-like functionality that would work with hypothetical undroppable types.

**Comment by anko** [comment]
Concern: Why `-> [T; 2]` over `-> (T, T)`?

I can't find precedent in `std` for either, so this may become the precedent that future functions returning "a pair of `T`s" would want to be consistent with.

- - -

## Informal survey

- Searching for occurrences of the alternatives in _this repo_:

  - [`"-> [T; 2]"`](https://github.com/search?q=repo%3Arust-lang%2Frust+%22-%3E+%5BT%3B+2%5D%22&type=code) → __1__ file: the current implementations of these functions.
  - [`"-> (T, T)"`](https://github.com/search?q=repo%3Arust-lang%2Frust+%22-%3E+%28T%2C+T%29%22&type=code) → __19__ files. They are all in tests; not exposed API. Only one appears to be a test for tuples specifically; the rest are "spontaneously arising".

- Searching for occurrences of the alternatives _everywhere on Github_, with Rust-only filter:

  - [`"-> [T; 2]"`](https://github.com/search?q=language%3Arust+%22-%3E+%5BT%3B+2%5D%22&type=code) → __326__ files.
  - [`"-> (T, T)"`](https://github.com/search?q=language%3Arust+%22-%3E+%28T%2C+T%29%22&type=code) → __2.8k__ files.

- I also [searched for `"fn min_max("`](https://github.com/search?q=language%3Arust+%22fn+min_max%28%22&type=code). In the 5 pages of results that Github will show me, I saw many implementations of similar functions returning a tuple of 2 items. I saw _no implementations_ returning an array of length 2.

- @timvermeulen's earlier similar PR (https://github.com/rust-lang/rust/pull/73857) also used `-> (T, T)`.

- - -

I believe these results indicate `-> (T, T)` is preferred, especially in this context.

By the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), I propose `-> (T, T)`.

**Comment by WaffleLapkin** [comment]
@anko as @scottmcm said, a lot of this is probably historic reasons — addition of matching / destructuring arrays is relatively new (*checks notes*, apparently it was stabilized in Rust 1.42, not *that* new).

I prefer and have used an array in those signatures, because I think it's nicer and better shows the intent.

It's nicer in the sense that arrays are more powerful than tuples — since they are homogenous they allow for more operations. Even though I think the result of this expression will almost always be destructured immediately, it's still nicer.

It better shows the intent in the sense that the methods return not just a pair of two values, but an ordered array. Also it highlightes the similarity to the `sort`.

Ultimately I don't think it matters much, but IMO the array is better here.

(ps. arrays and tuples also implement into/from)

**Comment by Ralith** [comment]
Looking forward to having this! Wanted it today when writing an algorithm that identifies the smaller of two values and walks it towards the greater.

**Comment by nmay231** [comment]
If the issue with #34162 takes longer than expected to resolve, could `minmax` (and potentially `minmax_by` since it's not affected) be stabilized before `minmax_by_key` is?

(Tbh, I'm not familiar with the requirements of stabilization)

**Comment by WaffleLapkin** [comment]
@nmay231 that is possible, we do partial stabilization all the time.

**Comment by Jules-Bertholet** [comment]
Could `minmax` and friends also be added to `Iterator`?

**Comment by notme4** [comment]
> It better shows the intent in the sense that the methods return not just a pair of two values, but an ordered array. 

I disagree, an array is a (possibly unordered) collection of a single kind of thing, whereas a tuple is an ordered collection of (possibly different kinds of) things, so a tuple shows intent better. Also min and max are opposites so they are arguably not even really the same kind of thing, making an array an even weaker option.

That being said the ordering of min and max is arbitrary and they have clear names, so returning a struct could be even clearer. The problem with that being then you have a struct cluttering the namespace.

**Comment by bend-n** [comment]
@notme4 how are tuples any more ordered than arrays?

**Comment by George-Ogden** [comment]
I think there's a bug/surprising feature in `cmp::minmax`:
It requires `Ord` trait but uses the `<=` operator, which uses the `PartialOrd::partial_cmp` method, instead of the `Ord::cmp` method. This is perhaps an issue with the `<=` operator, but it makes for some surprising results.

**Comment by GoldsteinE** [comment]
`Ord` disagreeing with `PartialOrd` is a logic error. There should be no surprising results for correct implementations.

**Comment by George-Ogden** [comment]
This assumes that everyone's implementation is always correct. If `Ord` and `PartialOrd` were to disagree by mistake, `cmp::minmax` causing an error would encourage you to check the `Ord` trait, not the `PartialOrd` trait. And the `#derive`'d implementation of `PartialOrd` doesn't default to using `Ord` when available, which only makes this issue more prevalent.

**Comment by hanna-kruppe** [comment]
Lots of APIs require an `Ord` bound but use the comparison operators internally because whatever they’re doing doesn’t make sense for partial orders, but it’s more convenient or more efficient to write them in terms of a particular binary comparison. This is not a `minmax` specific issue.

Yes, debugging cases where the traits are implemented inconsistently is a pain. But it’s better to catch that at the source with lints like `clippy::derive_ord_xor_partial_ord` and `clippy::derived_hash_with_manual_eq` than to complicate all code that tacitly relies on the contract of these traits. Those lints are already deny by default in clippy but not everyone uses clippy. Perhaps uplifting them is a good idea.

**Comment by crumblingstatue** [comment]
> Also, `minmax` has a [different meaning](https://en.wikipedia.org/wiki/Minimax).

Not in a way that would conflict with any other potential standard library method. 
In my opinion, you can't really be more clear about returning `(minimum, maximum)` than `minmax`, or `min_max` (feel free to bikeshed).
